src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -261,27 +261,45 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
-    } else if (mask_type->hi_as_long() < -1L) {
-      mask_max_bw = max_bw - 1;
-    } else {
-    // Case 3) Mask value range only includes +ve values.
-      assert(mask_type->lo_as_long() >= 0, "");
-      jlong clz = count_leading_zeros(mask_type->hi_as_long());
-      clz = bt == T_INT ? clz - 32 : clz;
-      mask_max_bw = max_bw - clz;
-    }
     if ( opc == Op_CompressBits) {
-      lo = mask_max_bw == max_bw ? lo : 0L;
-      // Compress operation is inherently an unsigned operation and
-      // result value range is primarily dependent on true count
-      // of participating mask value.
-      hi = mask_max_bw < max_bw ? (1L << mask_max_bw) - 1 : src_type->hi_as_long();
+      // Pattern: Integer/Long.compress(src_type, mask_type)
+      int result_bit_width;
+      int mask_bit_width = bt == T_INT ? 32 : 64;
+      if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
+        // Case 1) Mask value range includes -1, this negates the possibility of
+        // strictly non-negative result value range.
+        result_bit_width = mask_bit_width;
+      } else if (mask_type->hi_as_long() < -1L) {
+        // Case 2) Mask value range is less than -1, this indicates presence of at least
+        // one zero bit in the mask value, thereby constraining the result of compression
+        // to a +ve value range.
+        result_bit_width = mask_bit_width - 1;
+      } else {
+        // Case 3) Mask value range only includes +ve values, thus we can
+        // identify leading known zero bits of mask value and use this to
+        // constrain upper bound of result value range.
+        assert(mask_type->lo_as_long() >= 0, "");
+        jlong clz = count_leading_zeros(mask_type->hi_as_long());
+        // Here, result of clz is w.r.t to long argument, hence for integer argument
+        // we explicitly subtract 32 from the result.
+        clz = bt == T_INT ? clz - 32 : clz;
+        result_bit_width = mask_bit_width - clz;
+      }
+      // If number of bits required to accommodated mask value range is less than the
+      // full bit width of integral type, then MSB bit is guaranteed to be zero, thus
+      // compression result will never be a -ve value and we can safely set the
+      // lower bound of the result value range to zero.
+      lo = result_bit_width == mask_bit_width ? lo : 0L;
+
+      assert(hi == (bt == T_INT) ? max_jint : max_jlong, "");
+      assert((lo == (bt == T_INT) ? min_jint : min_jlong) || lo == 0, "");
+
+      // Following rules applies to upper bound estimation of results value range
+      // res.hi = src.hi iff src.hi > 0 else max_value
+      // if result_bit_width < mask_bit_width, then we can further constrain res.hi as follows.
+      // res.hi = MIN(res.hi, (1L << result_bit_width) - 1)
+      hi = src_type->hi_as_long() >= 0 ? src_type->hi_as_long() : hi;
+      hi = result_bit_width < mask_bit_width ? MIN2((jlong)((1L << result_bit_width) - 1L), hi) : hi;
```

Comment:
Note that if `result_bit_width == 63`, this computation will do `min_jlong - 1` which is UB.

Reply:
Thanks @merykitty , I liked your to the point an crisp comment.